sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1044,6 +1044,34 @@ class Analyzer(
           failOnOuterReference(p)
           p
       }
+
+      // SPARK-17348
+      // Looking for a potential incorrect result case.
+      // When a correlated predicate is a non-equality predicate
+      // it must be placed at the immediate child operator.
+      // Otherwise, the pull up of the correlated predicate
+      // will generate a plan with a different semantics
+      // which could return incorrect result.
+      var continue : Boolean = true
+      for (pm <- predicateMap if continue) {
+        assert(pm._2.nonEmpty, "Correlated predicate(s) does not exist.")
+        for (p <- pm._2 if continue)
+          p match {
+            case EqualTo(_, _) | EqualNullSafe(_, _) =>
+              None
+            case _ =>
+              assert(transformed.children.nonEmpty)
+              if (!(transformed.isInstanceOf[Project]) ||
```

Comment:
@hvanhovell The key point of this PR is on the second condition, not the check for the `Project`. Its intention is if the correlated predicate is not at the immediate child, we are about to pull the correlated predicate up through a number of operations. It is very hard to prove that which operations we can pull the predicate through without changing the meaning of the original SQL. Yes, there are cases that it may regress, like the example @srinathshankar posted.

I do not understand your second comment on putting in the previous block. Could you please explain more?